File: src/report.js
'use strict';

// The report.
//
// This tool would have no standing if its own output failed the rules it
// enforces, so: colour only ever reinforces a word that already says the same
// thing, NO_COLOR and TERM=dumb are honoured, nothing is dimmed, nothing is
// redrawn in place, and everything wraps to the real terminal width.

const SEVERITY_WEIGHT = { critical: 25, serious: 10, moderate: 4, minor: 1 };
const SEVERITY_ORDER = ['critical', 'serious', 'moderate', 'minor'];

const DIMENSIONS = {
  vision: 'Vision and screen readers',
  motor: 'Interaction and input',
  clarity: 'Comprehension',
};

function styler(enabled) {
  const wrap = (code) => (s) => (enabled ? `\x1b[${code}m${s}\x1b[0m` : String(s));
  return {
    bold: wrap(1),
    red: wrap(91),
    yellow: wrap(93),
    green: wrap(92),
    cyan: wrap(96),
    // Deliberately not a "dim" helper. Faint text is one of the things this
    // tool reports as a fault; it has no business using it.
    plain: (s) => String(s),
  };
}

function colourEnabled(opts) {
  if (opts.color === false) return false;
  if (opts.color === true) return true;
  if (process.env.NO_COLOR) return false;
  if (process.env.TERM === 'dumb') return false;
  return Boolean(process.stdout.isTTY);
}

function wrapText(text, width, indent = '') {
  const out = [];
  for (const para of String(text).split('\n')) {
    let line = indent;
    for (const word of para.split(/\s+/).filter(Boolean)) {
      if (line.length > indent.length && line.length + 1 + word.length > width) {
        out.push(line);
        line = indent + word;
      } else {
        line = line.length > indent.length ? `${line} ${word}` : indent + word;
      }
    }
    out.push(line);
  }
  return out;
}

function score(findings) {
  const penalty = findings
    .filter((f) => f.status === 'fail')
    .reduce((a, f) => a + (SEVERITY_WEIGHT[f.severity] || 0), 0);
  return Math.max(0, 100 - penalty);
}

function grade(n) {
  if (n >= 90) return 'good';
  if (n >= 70) return 'workable, with gaps';
  if (n >= 45) return 'hard going';
  return 'largely unusable';
}

function summarise(findings) {
  const counts = { critical: 0, serious: 0, moderate: 0, minor: 0, pass: 0, skip: 0 };
  for (const f of findings) {
    if (f.status === 'pass') counts.pass++;
    else if (f.status === 'skip') counts.skip++;
    else counts[f.severity] = (counts[f.severity] || 0) + 1;
  }
  return counts;
}

/** Does anything reach the threshold that should fail a build? */
function shouldFail(findings, level) {
  if (level === 'none') return false;
  const cut = SEVERITY_ORDER.indexOf(level);
  if (cut < 0) return false;
  return findings.some((f) => f.status === 'fail'
    && SEVERITY_ORDER.indexOf(f.severity) <= cut);
}

function render(result, opts = {}) {
  const { findings, target, meta } = result;
  const colour = colourEnabled(opts);
  const c = styler(colour);
  const width = Math.max(40, Math.min(opts.width || process.stdout.columns || 80, 100));
  const lines = [];
  const counts = summarise(findings);
  const total = score(findings);

  const label = {
    critical: c.red('CRITICAL'),
    serious: c.red('SERIOUS '),
    moderate: c.yellow('MODERATE'),
    minor: c.yellow('MINOR   '),
  };

  lines.push('');
  lines.push(c.bold(`Accessibility report for: ${target.cmd} ${target.args.join(' ')}`.trim()));
  lines.push('');
  lines.push(`Score ${c.bold(`${total}/100`)} — ${grade(total)}`);
  lines.push(`${counts.critical} critical, ${counts.serious} serious, ${counts.moderate} moderate, `
    + `${counts.minor} minor, ${counts.pass} passed, ${counts.skip} not applicable`);
  if (!meta.pty) {
    lines.push('');
    lines.push(...wrapText('Note: no usable script(1) on this machine, so every terminal probe was '
      + 'skipped. Only the piped behaviour was graded, which misses most colour and '
      + 'animation faults.', width));
  }

  for (const [dim, title] of Object.entries(DIMENSIONS)) {
    const group = findings.filter((f) => f.dimension === dim);
    if (!group.length) continue;
    const fails = group.filter((f) => f.status === 'fail');
    const dimScore = score(group);

    lines.push('');
    lines.push(c.bold(title) + `  —  ${fails.length} of ${group.length} checks failed, score ${dimScore}/100`);
    lines.push('-'.repeat(Math.min(width, 72)));

    const ordered = [
      ...SEVERITY_ORDER.flatMap((s) => fails.filter((f) => f.severity === s)),
      ...(opts.quiet ? [] : group.filter((f) => f.status !== 'fail')),
    ];

    let separated = false;
    for (const f of ordered) {
      if (f.status === 'fail') {
        lines.push('');
        lines.push(`${label[f.severity] || f.severity}  ${c.bold(f.title)}  [${f.id}]`);
        lines.push(...wrapText(f.detail, width, '          '));
        for (const e of f.evidence || []) {
          lines.push(...wrapText(`· ${e}`, width, '          '));
        }
        if (f.remedy) {
          lines.push(...wrapText(`${c.cyan('Fix:')} ${f.remedy}`, width, '          '));
        }
      } else {
        // One blank line separates the failures from the roll-call of what
        // passed, so the two never run together when read aloud.
        if (!separated) { lines.push(''); separated = true; }
        if (f.status === 'pass') {
          lines.push(`${c.green('PASS')}      ${f.title}`);
        } else {
          const [first, ...more] = wrapText(`${f.title} — ${f.detail}`, width - 10);
          lines.push(`N/A       ${first}`);
          more.forEach((l) => lines.push(`          ${l}`));
        }
      }
    }
  }

  lines.push('');
  return lines.join('\n');
}

function toJson(result) {
  const { findings, target, meta } = result;
  return JSON.stringify({
    target,
    score: score(findings),
    grade: grade(score(findings)),
    counts: summarise(findings),
    dimensions: Object.fromEntries(Object.keys(DIMENSIONS).map((d) => {
      const g = findings.filter((f) => f.dimension === d);
      return [d, { score: score(g), checks: g.length, failed: g.filter((f) => f.status === 'fail').length }];
    })),
    meta: { pty: meta.pty, probes: meta.probeCount, durationMs: meta.durationMs },
    findings: findings.map((f) => ({
      id: f.id,
      dimension: f.dimension,
      status: f.status,
      severity: f.severity,
      title: f.title,
      detail: f.detail,
      evidence: f.evidence,
      remedy: f.remedy,
    })),
  }, null, 2);
}

module.exports = {
  render, toJson, score, grade, summarise, shouldFail,
  SEVERITY_ORDER, SEVERITY_WEIGHT, DIMENSIONS,
};

File: internal/report/report.go
// Package report renders an audit.
//
// This tool would have no standing if its own output failed the rules it
// enforces, so: colour only ever reinforces a word that already says the same
// thing, NO_COLOR and TERM=dumb are honoured, nothing is dimmed, nothing is
// redrawn in place, and everything wraps to the real terminal width.
package report

import (
	"encoding/json"
	"fmt"
	"os"
	"strings"

	"github.com/Nessaek/cli-a11y/internal/check"
	"github.com/Nessaek/cli-a11y/internal/probe"
)

// Dimension titles, in the order they are shown.
var dimensions = []struct{ key, title string }{
	{check.Vision, "Vision and screen readers"},
	{check.Motor, "Interaction and input"},
	{check.Clarity, "Comprehension"},
}

var severityOrder = []check.Severity{check.Critical, check.Serious, check.Moderate, check.Minor}

// Result is a finished audit.
type Result struct {
	Target   probe.Target
	Findings []check.Finding
	Meta     probe.Meta
	PTY      bool
}

// Options controls rendering.
type Options struct {
	Colour *bool // nil means decide from the environment
	Quiet  bool
	Width  int
}

// Score is the audit score out of 100.
func Score(fs []check.Finding) int {
	penalty := 0
	for _, f := range fs {
		if f.Failed() {
			penalty += f.Sev().Weight()
		}
	}
	return max(0, 100-penalty)
}

// Grade puts a score in words.
func Grade(n int) string {
	switch {
	case n >= 90:
		return "good"
	case n >= 70:
		return "workable, with gaps"
	case n >= 45:
		return "hard going"
	default:
		return "largely unusable"
	}
}

// Counts tallies findings by outcome.
type Counts struct {
	Critical, Serious, Moderate, Minor, Pass, Skip int
}

// Summarise counts the findings.
func Summarise(fs []check.Finding) Counts {
	var c Counts
	for _, f := range fs {
		switch {
		case f.Passed():
			c.Pass++
		case !f.Failed():
			c.Skip++
		default:
			switch f.Sev() {
			case check.Critical:
				c.Critical++
			case check.Serious:
				c.Serious++
			case check.Moderate:
				c.Moderate++
			case check.Minor:
				c.Minor++
			}
		}
	}
	return c
}

// ShouldFail reports whether anything reaches the threshold that fails a build.
func ShouldFail(fs []check.Finding, level check.Severity) bool {
	if level == check.None {
		return false
	}
	for _, f := range fs {
		if f.Failed() && f.Sev() <= level {
			return true
		}
	}
	return false
}

type styler struct{ on bool }

func (s styler) wrap(code string, v string) string {
	if !s.on {
		return v
	}
	return "\x1b[" + code + "m" + v + "\x1b[0m"
}

func (s styler) bold(v string) string  { return s.wrap("1", v) }
func (s styler) red(v string) string   { return s.wrap("91", v) }
func (s styler) amber(v string) string { return s.wrap("93", v) }
func (s styler) green(v string) string { return s.wrap("92", v) }
func (s styler) cyan(v string) string  { return s.wrap("96", v) }

// Deliberately no dim helper. Faint text is one of the things this tool reports
// as a fault; it has no business using it.

func colourEnabled(o Options) bool {
	if o.Colour != nil {
		return *o.Colour
	}
	if os.Getenv("NO_COLOR") != "" || os.Getenv("TERM") == "dumb" {
		return false
	}
	fi, err := os.Stdout.Stat()
	return err == nil && fi.Mode()&os.ModeCharDevice != 0
}

func wrapText(text string, width int, indent string) []string {
	var out []string
	for _, para := range strings.Split(text, "\n") {
		line := indent
		for _, word := range strings.Fields(para) {
			switch {
			case len(line) > len(indent) && len(line)+1+len(word) > width:
				out = append(out, line)
				line = indent + word
			case len(line) > len(indent):
				line += " " + word
			default:
				line = indent + word
			}
		}
		out = append(out, line)
	}
	return out
}

// Render produces the terminal report.
func Render(r Result, o Options) string {
	c := styler{colourEnabled(o)}
	width := o.Width
	if width == 0 {
		width = 80
	}
	width = max(40, min(width, 100))

	counts := Summarise(r.Findings)
	total := Score(r.Findings)
	labels := map[check.Severity]string{
		check.Critical: c.red("CRITICAL"),
		check.Serious:  c.red("SERIOUS "),
		check.Moderate: c.amber("MODERATE"),
		check.Minor:    c.amber("MINOR   "),
	}

	var b strings.Builder
	line := func(s string) { b.WriteString(s + "\n") }
	lines := func(ss []string) {
		for _, s := range ss {
			line(s)
		}
	}

	target := strings.TrimSpace(r.Target.Cmd + " " + strings.Join(r.Target.Args, " "))
	line("")
	line(c.bold("Accessibility report for: " + target))
	line("")
	line(fmt.Sprintf("Score %s — %s", c.bold(fmt.Sprintf("%d/100", total)), Grade(total)))
	line(fmt.Sprintf("%d critical, %d serious, %d moderate, %d minor, %d passed, %d not applicable",
		counts.Critical, counts.Serious, counts.Moderate, counts.Minor, counts.Pass, counts.Skip))
	if !r.PTY {
		line("")
		lines(wrapText("Note: no pseudo-terminal could be opened, so every terminal probe was skipped. "+
			"Only piped behaviour was graded, which misses most colour and animation faults.", width, ""))
	}

	for _, d := range dimensions {
		var group []check.Finding
		for _, f := range r.Findings {
			if f.Dimension == d.key {
				group = append(group, f)
			}
		}
		if len(group) == 0 {
			continue
		}
		failed := 0
		for _, f := range group {
			if f.Failed() {
				failed++
			}
		}
		line("")
		line(c.bold(d.title) + fmt.Sprintf("  —  %d of %d checks failed, score %d/100", failed, len(group), Score(group)))
		line(strings.Repeat("-", min(width, 72)))

		var ordered []check.Finding
		for _, sev := range severityOrder {
			for _, f := range group {
				if f.Failed() && f.Sev() == sev {
					ordered = append(ordered, f)
				}
			}
		}
		if !o.Quiet {
			for _, f := range group {
				if !f.Failed() {
					ordered = append(ordered, f)
				}
			}
		}

		separated := false
		for _, f := range ordered {
			if f.Failed() {
				line("")
				line(fmt.Sprintf("%s  %s  [%s]", labels[f.Sev()], c.bold(f.Title), f.ID))
				lines(wrapText(f.Detail, width, "          "))
				for _, e := range f.Evidence {
					lines(wrapText("· "+e, width, "          "))
				}
				if f.Remedy != "" {
					lines(wrapText(c.cyan("Fix:")+" "+f.Remedy, width, "          "))
				}
				continue
			}
			// One blank line separates the failures from the roll-call of what
			// passed, so the two never run together when read aloud.
			if !separated {
				line("")
				separated = true
			}
			if f.Passed() {
				line(c.green("PASS") + "      " + f.Title)
				continue
			}
			wrapped := wrapText(f.Title+" — "+f.Detail, width-10, "")
			line("N/A       " + wrapped[0])
			for _, w := range wrapped[1:] {
				line("          " + w)
			}
		}
	}

	line("")
	return b.String()
}

type jsonReport struct {
	Target     map[string]any            `json:"target"`
	Score      int                       `json:"score"`
	Grade      string                    `json:"grade"`
	Counts     Counts                    `json:"counts"`
	Dimensions map[string]map[string]int `json:"dimensions"`
	Meta       map[string]any            `json:"meta"`
	Findings   []check.Finding           `json:"findings"`
}

// JSON renders the machine-readable report.
func JSON(r Result) (string, error) {
	dims := map[string]map[string]int{}
	for _, d := range dimensions {
		var group []check.Finding
		failed := 0
		for _, f := range r.Findings {
			if f.Dimension == d.key {
				group = append(group, f)
				if f.Failed() {
					failed++
				}
			}
		}
		dims[d.key] = map[string]int{"score": Score(group), "checks": len(group), "failed": failed}
	}
	out := jsonReport{
		Target:     map[string]any{"cmd": r.Target.Cmd, "args": r.Target.Args},
		Score:      Score(r.Findings),
		Grade:      Grade(Score(r.Findings)),
		Counts:     Summarise(r.Findings),
		Dimensions: dims,
		Meta: map[string]any{
			"pty":        r.PTY,
			"probes":     r.Meta.ProbeCount,
			"durationMs": r.Meta.Duration.Milliseconds(),
		},
		Findings: r.Findings,
	}
	b, err := json.MarshalIndent(out, "", "  ")
	return string(b), err
}
